pkg/middileware/middleware.go

```diff
@@ -25,15 +25,89 @@ SOFTWARE.
 package middleware
 
 import (
+	"net/http"
+
+	"github.com/praromvik/praromvik/models"
+	"github.com/praromvik/praromvik/pkg/auth"
+	"github.com/praromvik/praromvik/pkg/error"
+
 	"github.com/go-chi/chi/v5"
 	"github.com/go-chi/chi/v5/middleware"
 )
 
 func AddMiddlewares(router *chi.Mux) {
-	router.Use(middleware.RequestID)
 	router.Use(middleware.Logger)
+	router.Use(middleware.RequestID)
 	router.Use(middleware.Recoverer)
 	router.Use(middleware.URLFormat)
-	router.Use(middleware.RequestID)
-	router.Use(middleware.Logger)
+}
+
+func SecurityMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		valid, err := auth.SessionValid(r)
+		if err != nil {
+			error.HandleError(w, http.StatusUnauthorized, "Failed to validate session: "+err.Error(), err)
+			return
+		}
+		if !valid {
+			error.HandleError(w, http.StatusUnauthorized, "Invalid session token", err)
+			return
+		}
+		authenticated, err := auth.IsAuthenticated(r)
+		if err != nil {
+			error.HandleError(w, http.StatusUnauthorized, "Failed to check authentication status: "+err.Error(), err)
+			return
+		}
+
+		if !authenticated {
+			error.HandleError(w, http.StatusUnauthorized, "Unauthenticated session token", err)
+			return
+		}
+		next.ServeHTTP(w, r)
+	})
+}
+
+func AdminAccess(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {
+		roleType, err := auth.GetSessionRole(request)
+		if err != nil {
+			error.HandleError(writer, http.StatusUnauthorized, "Failed to retrieve role from session", err)
+			return
+		}
+		if roleType != models.Admin {
+			error.HandleError(writer, http.StatusUnauthorized, "Insufficient privileges. Admin or Moderator access required", err)
+			return
+		}
+		next.ServeHTTP(writer, request)
+	})
+}
+
+func AdminOrModeratorAccess(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {
+		roleType, err := auth.GetSessionRole(request)
+		if err != nil {
+			error.HandleError(writer, http.StatusUnauthorized, "Failed to retrieve role from session", err)
+			return
+		}
+		if !(roleType == models.Admin || roleType == models.Moderator) {
+			error.HandleError(writer, http.StatusUnauthorized, "Insufficient privileges. Admin or Moderator access required", err)
+			return
+		}
```

Comment:
We can make a common function, with roles as param. & call that from each of these funcs.